### ISSUE TITLE
More NaNs

### ISSUE THREAD

**Issue by lars-nieradzik** [frgfm/torch-cam#186]
### Bug description

Thank you for fixing the previous NaN bug. This solves most issues. However, sometimes these NaNs still occur. With the newest commit, the tensors are zero.

I would suggest adding to torch.nansum a warning message that informs the user that NaN was found. Otherwise, the activation map is just zero and the user is not aware that a problem occurred. I am referring to line https://github.com/frgfm/torch-cam/blob/0a5e06051440e27de6027ec382517a2c71686298/torchcam/methods/core.py#L190

Here an example from ImageNet that produces NaN: ILSVRC2012_val_00033712.JPEG

![ILSVRC2012_val_00033712](https://user-images.githubusercontent.com/102812109/190910494-ce23665d-fe43-4a44-8d6d-23c7b464c3be.JPEG)

### Code snippet to reproduce the bug

```python
from torchvision.io.image import read_image
from torchvision.transforms.functional import normalize, resize, to_pil_image
import torch
import timm
from torchvision.models import resnet50
from torchcam.methods import *

model = timm.create_model("efficientnet_b0", pretrained=True).eval()
cam_extractor = ScoreCAM(model, target_layer="blocks.6")
# Get your input
img = read_image("ILSVRC2012_val_00033712.JPEG")
# Preprocess it for your chosen model
input_tensor = normalize(resize(img, (224, 224)) / 255., [0.485, 0.456, 0.406], [0.229, 0.224, 0.225])

# Preprocess your data and feed it to the model
out = model(input_tensor.unsqueeze(0))
# Retrieve the CAM by passing the class index and the model output
activation_map = cam_extractor(out.squeeze(0).argmax().item(), out)

print(activation_map)
```

### Error traceback

```python
[tensor([[[0., 0., 0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0., 0., 0.]]])]
```

### Environment

not relevant

**Comment by frgfm** [comment]
Hey again @lars-nieradzik :wave:

Thanks for the issue, I appreciate how specific you are, it's much easier to debug than most requests I get :pray: 
On that part, I have doubts that might be the best solution. If NaNs/zero arise because of numerization / float precision, they need to be handled by TorchCAM. However, if for some reasons, the produced feature map is uniform spatially the CAM is basically zero: there is no special spatial region that influenced the model output. And in that case, I'd argue it's quite normal that the output CAM is only zeros.

About the `torch.nansum`, there are only two cases:
- NaNs come from TorchCAM processing errors (mistake in implementations, division by zeros, etc.): those need to be handled
- NaNs come from the model / input: if that's the case, perhaps yes we could throw a warning, but this is extending the "responsibility scope" of TorchCAM beyond its primary use

I'll run some tests on the image you sent, but feel free to open a PR if you have a specific idea in mind :)

**Comment by frgfm** [comment]
Investigating a bit more, on the image you sent, I located the source of zeros:
- the cam is passed through a ReLU before being normalized
- however, in this example the raw cam is negative

So for this example, I'd argue that there is no error: there is no positive correlation of spatial features on that feature map with the output score (because all values are negative). What do you think?

**Comment by lars-nieradzik** [comment]
Thank you for your quick answers. Regarding the image, you are right. I just assumed that the zero activation map happened due to torch.nansum(). If all the NaN errors all occur naturally (input has NaN etc.), then I guess a debug msg would not be necessary.

However, I am still seeing NaN in other images. I added `if torch.isnan(weight).any(): print("ERROR")`  to core.py (compute_cams) and ran the library again on a few hundred images.

![ILSVRC2012_val_00029594](https://user-images.githubusercontent.com/102812109/190916673-d6242d77-0592-405d-9cee-900c8cecd0d0.JPEG)

```python
from torchvision.io.image import read_image
from torchvision.transforms.functional import normalize, resize, to_pil_image
import torch
import timm
from torchvision.models import resnet50
from torchcam.methods import *

model = timm.create_model("densenet121", pretrained=True).eval()
cam_extractor = SmoothGradCAMpp(model, target_layer="features.denseblock4.denselayer16.conv2")
# Get your input
img = read_image("ILSVRC2012_val_00029594.JPEG")
# Preprocess it for your chosen model
input_tensor = normalize(resize(img, (224, 224)) / 255., [0.485, 0.456, 0.406], [0.229, 0.224, 0.225])

# Preprocess your data and feed it to the model
out = model(input_tensor.unsqueeze(0))
# Retrieve the CAM by passing the class index and the model output
activation_map = cam_extractor(out.squeeze(0).argmax().item(), out)

print(activation_map)
```

**Comment by frgfm** [comment]
With the script you just shared and the image, I'm getting a decent activation map:
```
[tensor([[[1.8036e-02, 0.0000e+00, 3.9917e-04, 0.0000e+00, 0.0000e+00,
          0.0000e+00, 0.0000e+00],
         [2.6609e-02, 4.7441e-02, 8.1286e-02, 6.2826e-02, 6.9473e-02,
          2.0333e-02, 1.4543e-02],
         [3.5339e-02, 1.3377e-01, 5.7993e-01, 6.9856e-01, 5.0984e-01,
          2.2307e-01, 8.1495e-02],
         [3.8667e-02, 3.7683e-01, 8.3038e-01, 1.0000e+00, 7.8089e-01,
          3.1000e-01, 1.0363e-01],
         [8.2719e-02, 2.2919e-01, 6.9044e-01, 8.6880e-01, 6.2525e-01,
          4.4371e-01, 2.1152e-01],
         [3.7904e-02, 1.0292e-01, 2.5570e-01, 3.0357e-01, 6.2674e-01,
          6.0119e-01, 4.8994e-01],
         [0.0000e+00, 3.9352e-02, 1.9601e-01, 3.2979e-01, 5.8944e-01,
          5.9874e-01, 4.2369e-01]]])]

```

Regarding the NaNs for gradient methods, this is the same problem as normalization for ScoreCAMs previously (division by zero), that's for spotting this. I just opened a PR #187 to add an eps to the denominator, that should do it :)

**Comment by frgfm** [comment]
@lars-nieradzik feel free to reopen if you encounter any more problems :+1:

**Comment by almeidava93** [comment]
Hello, guys! I'm having the same problem here. I'm going to share my code. I'm using a model I trained. 

```
`from torchcam.methods import SmoothGradCAMpp, LayerCAM
from torchcam.utils import overlay_mask

idx = random.randint(0, dataset_sizes['train'])
img = train_set[idx][0]
label = train_set[idx][1]

# data visualization
fig, axs = plt.subplots(3, 1, figsize=(5, 10), tight_layout=True)
axs[0].set_title(f"{'X ray with pneumonia' if label == 1 else 'X ray without pneumonia'}")
axs[0].axis('off')
axs[0].imshow(np.transpose(vutils.make_grid(img, padding=2, normalize=True).cpu(),(1,2,0)))

cam_extractor = SmoothGradCAMpp(model, input_shape=(3, 256, 256))

# Preprocess your data and feed it to the model
out = model(img.unsqueeze(0)) # unsqueeze is because the model expects 4d tensor

# Retrieve the CAM by passing the class index and the model output
cams = cam_extractor(out.squeeze(0).argmax().item(), out)

# The raw CAM
for name, cam in zip(cam_extractor.target_names, cams):
  axs[1].set_title("heatmap")
  axs[1].axis('off')
  axs[1].imshow(cam.squeeze(0).cpu().numpy())

# Overlayed on the image
for name, cam in zip(cam_extractor.target_names, cams):
  result = overlay_mask(to_pil_image(img, mode='RGB'), to_pil_image(cam.squeeze(0), mode='F'), alpha=0.5)
  axs[2].set_title("Grad-CAM overlayed heatmap")
  axs[2].axis('off')
  axs[2].imshow(result)
  break

plt.show()

# Once you're finished, clear the hooks on your model
cam_extractor.remove_hooks()
```

When I inspect values of cam I get:
```
tensor([[[nan, nan, nan, nan, nan, nan, nan, nan],
         [nan, nan, nan, nan, nan, nan, nan, nan],
         [nan, nan, nan, nan, nan, nan, nan, nan],
         [nan, nan, nan, nan, nan, nan, nan, nan],
         [nan, nan, nan, nan, nan, nan, nan, nan],
         [nan, nan, nan, nan, nan, nan, nan, nan],
         [nan, nan, nan, nan, nan, nan, nan, nan],
         [nan, nan, nan, nan, nan, nan, nan, nan]]], device='cuda:0')
```

**Comment by frgfm** [comment]
Hi @almeidava93,

I suggest opening another issue or discussion as this is most likely not related to the reason in this topic. Only one check to do before: are you using the dev updated version? Or the last stable version?

And for me to help, I'd need to be able to reproduce but I cannot run your snippet (missing some object definitions and the input data I imagine). Could you try to come up with a minimal reproducable snippet? (if the only information we have is that it only happens for some images in your dataset and we don't have them or the models weights, I can only assume/guess but this is far less efficient than if I had them :sweat_smile: )

Best
